tools/pipelines/build-docs.yml

```diff
@@ -100,27 +100,36 @@ resources:
   pipelines:
   - pipeline: common_definitions
     source: Build - common-definitions
+    branch: $(Build.SourceBranchName)
```

Comment:
Have we seen the expected result with the change as is? I would have expected that we needed "compile-time" syntax for the variables, i.e. `${{ variables['Build.SourceBranchName'] }}`.

Reply:
I did some testing on a test branch and it seems like both `${{ variables['Build.SourceBranchName'] }}` and `$(Build.SourceBranchName` causes the resource to not be found. However, after changing the branch to target main explicitly, the branch works as expected: https://dev.azure.com/fluidframework/internal/_build/results?buildId=229490&view=logs&j=897f79ef-2c90-5d46-29c0-b0ed22a2d39c&t=55b819c0-adcd-5b13-edef-a210850679b8
in this example, the pipeline downloads from Build-azure main instead of the latest run which was from lts.

Based on @Josmithr comment, would it be ok to pass in `main` explicitly?

Reply:
Not sure about Josh's point, but on second thought I think what we actually need to do to get this working for several branches (assuming it should work for several branches), is update all the downloading of artifacts from each resource [here](https://github.com/alexvy86/FluidFramework/blob/a33982b1891c42a50405d1e238773f721117f6a4/tools/pipelines/templates/upload-json-steps.yml#L22-L39), to do what the DownloadPipelineArtifact task does in other places like [here](https://github.com/alexvy86/FluidFramework/blob/a33982b1891c42a50405d1e238773f721117f6a4/tools/pipelines/templates/include-test-real-service.yml#L165-L182). Changing the branch in the resource might be ok if this pipeline will actually only ever run for a specific branch; updating the artifact download steps gives us more flexibility.

Tangent: we seem to only have that one use of the `upload-json-steps` template so I'm going to create an ADO task to inline it. It refers to the resource names from the pipeline so they're already extremely co-dependent, and there's other risks of having it be a separate template file.

Reply:
That makes sense, I'll take another look into why the `Build.SourceBranchName` syntax wasn't working. Also, I'll add the branch name variable to the download task as well.

Reply:
I think I know why it wasn't working; the `resources` section of a pipeline definition probably just doesn't support using variables at all. It's something that ADO needs to understand "very early" during the execution of a pipeline, probably before it even has a chance to start doing variable replacements. The runtime syntax (`$()`) I'm almost positive wouldn't work; the compile-time one I felt more confident about, but wouldn't be entirely surprised if we can't use them in that particular section of the file.

Reply:
And thus ADO might have been looking for a branch literally named `"$(Build.SourceBranchName)"` or `"${{ variables['Build.SourceBranchName'] }}"`.

Or, maybe the compile-time syntax works but we need to use `${{ variables['Build.SourceBranch'] }}` (SourceBranch instead of SourceBranchName; one is a more complete "pointer"; [docs](https://learn.microsoft.com/en-us/azure/devops/pipelines/build/variables?view=azure-devops&tabs=yaml#build-variables-devops-services)).

Reply:
I wonder if ADO offers any sort of linter to detect unsupported syntax. It would be great if we could get earlier feedback on these sorts of things than "either the pipeline works, or it doesn't, or it does mysterious things instead of what I wanted".

Reply:
Ooh, well, there is at least this: https://marketplace.visualstudio.com/items?itemName=ms-azure-devops.azure-pipelines

Not sure how comprehensive it is, but it sure seems worth a look. I'm going to try it out.